copr: introduce rpn_fn macro to generate RPN function code under the new framework

sticnarf opened the pull request:
<!--
Thank you for contributing to TiKV! Please read TiKV's [CONTRIBUTING](https://github.com/tikv/tikv/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

## What have you changed? (mandatory)

#4859 shows what the code looks like. It is huge amount of code if we write it on our own.

This macro is an atrribute on a free function. People can use the generated `foo_fn()` function to get an `RpnFn`. See comments in `function.rs` for usage.

Check the test under `components/cop_codegen/rpn_functions.rs` to see what code the macro will generate.

## What are the type of the changes? (mandatory)

Engineering

## How has this PR been tested? (mandatory)

CI

## Does this PR affect documentation (docs) or release note? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

kennytm commented on components/cop_codegen/src/lib.rs:
```diff
@@ -36,3 +43,19 @@ pub fn aggr_function_derive(input: TokenStream) -> TokenStream {
+        Ok(tokens) => TokenStream::from(tokens),
+        Err(e) => {
+            e.emit();
+            std::process::exit(1)
```
Could you return an empty TokenStream or something like that here?

sticnarf replied:
emm...What's the benefit of returning an empty TokenStream?

kennytm replied:
It allows the compiler collect as many errors as possible. See also https://github.com/rust-lang/rust/issues/54140#issuecomment-434969649

lonng commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,595 @@ impl RpnFunctionOpts {
+}
+
+fn ctx_type() -> TokenStream {
+    "crate::coprocessor::dag::expr::EvalContext"
```
```suggestion
    quote!(crate::coprocessor::dag::expr::EvalContext).into()
```

lonng commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,595 @@ impl RpnFunctionOpts {
+}
+
+fn payload_type() -> TokenStream {
+    "crate::coprocessor::dag::rpn_expr::types::RpnFnCallPayload"
```
Use `quote!` instead of `parse + unwrap`

lonng commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,595 @@ impl RpnFunctionOpts {
+}
+
+fn result_type() -> TokenStream {
+    "crate::coprocessor::Result<crate::coprocessor::codec::data_type::VectorValue>"
```
(ditto)

lonng commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,595 @@ impl RpnFunctionOpts {
+}
+
+fn rpn_fn_type() -> TokenStream {
+    "crate::coprocessor::dag::rpn_expr::function::RpnFn"
```
(ditto)

lonng commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,595 @@ impl RpnFunctionOpts {
+        let mut generics = self.item_fn.decl.generics.clone();
+        generics
+            .params
+            .push(parse_str(&format!("D_: {}", ARG_DEF)).unwrap());
```
I think `T: xxx` is more idiom.

sticnarf replied:
Uncommon name is used on purpose to avoid name confliction.

lonng commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,595 @@ impl RpnFunctionOpts {
+                    ctx: &mut #ctx_type,
+                    payload: #payload_type,
+                ) -> #result_type {
+                    panic!("Cannot apply {} on {:?}", #fn_name, self)
```
It's better to provide a default implementation of `eval` in `**_Fn` trait. `impl #impl_generics #fn_trait_ident #ty_generics for #tp_ident #where_clause {}` is adequate.

lonng replied:
In further we can put `impl #impl_generics #fn_trait_ident #ty_generics for #tp_ident #where_clause {}` in `generate_fn_trait` directly, and remove this function.

sticnarf replied:
No, we cannot do that. We need an implementation for all `ArgDef`. Otherwise, [this line](https://github.com/tikv/tikv/pull/4860/files/0aa2e6302a0d2da2653ae8a474d5f2e13d5e4abf#diff-3e003cbe17506910d113fa151fcad42bR251) won't compile.

lonng commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,595 @@ impl RpnFunctionOpts {
+                    let arg = &self;
+                     let mut result = Vec::with_capacity(rows);
+                     for row in 0..rows {
+                         #extract_args
```
We can remove L185-L196 and generate `extract_args` and `call_args` in `quote!` directly. See: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9f97453d8f5c10370b82d757bd8d9ace

sticnarf replied:
Good idea!

lonng commented on src/coprocessor/dag/rpn_expr/function.rs:
```diff
@@ -345,3 +385,171 @@ impl Helper {
 }
+
+#[derive(Clone, Copy)]
+/// A RPN function
```
```suggestion
/// An RPN function
```

lonng commented on src/coprocessor/dag/rpn_expr/function.rs:
```diff
@@ -1,6 +1,46 @@
 // Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
 
-use super::types::RpnFnCallPayload;
+//! People implementing RPN functions with fixed argument type and count don't necessarily
```
@kennytm Please help him to polish the comments.

kennytm replied:
......

lonng commented on src/coprocessor/dag/rpn_expr/impl_compare.rs:
```diff
@@ -22,15 +21,29 @@ impl<C: Comparer> RpnFnCompare<C> {
+        ctx: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue> {
+        let rpn_fn = compare_fn::<C>();
```
Is this a temporary code? Is need to call `compare_fn` every time?

sticnarf replied:
It's temporary. As `RpnFunction` trait is removed in #4861, this is also removed.

lonng commented:
Good job, 👍

breeswish commented on components/cop_codegen/Cargo.toml:
```diff
@@ -12,3 +12,4 @@ darling = "0.9"
 syn = "0.15"
 quote = "0.6"
 proc-macro2 = "0.4"
+heck = "0.3"
```
reserve a new line

breeswish commented on src/coprocessor/dag/rpn_expr/function.rs:
```diff
@@ -1,6 +1,46 @@
+//!     payload: RpnFnCallPayload<'_>
+//! ) -> Result<Option<Int>> {
+//!     // Your RPN function logic
+//! }
```
Please also add an example about writing a function that receives and handles raw vector/scalar arguments manually.

sticnarf replied:
Example added. Please take a look.

kennytm commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,591 @@ impl RpnFunctionOpts {
+            .skip(meta.len()) // ctx or payload are not function args
+            .map(parse_arg_type)
+            .collect::<Result<Vec<_>>>()?;
+        let camel_name = format!("{}", item_fn.ident).to_camel_case();
```
```suggestion
        let camel_name = item_fn.ident.to_string().to_camel_case();
```

kennytm commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,591 @@ impl RpnFunctionOpts {
+            .map(parse_arg_type)
+            .collect::<Result<Vec<_>>>()?;
+        let camel_name = format!("{}", item_fn.ident).to_camel_case();
+        let fn_trait_ident = parse_str(&format!("{}_Fn", camel_name)).unwrap();
```
```suggestion
        let fn_trait_ident = Ident::new(&format!("{}_Fn", camel_name), Span::call_site());
```

kennytm commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,591 @@ impl RpnFunctionOpts {
+            .collect::<Result<Vec<_>>>()?;
+        let camel_name = format!("{}", item_fn.ident).to_camel_case();
+        let fn_trait_ident = parse_str(&format!("{}_Fn", camel_name)).unwrap();
+        let evaluator_ident = parse_str(&format!("{}_Evaluator", camel_name)).unwrap();
```
```suggestion
        let evaluator_ident = Ident::new(&format!("{}_Evaluator", camel_name), Span::call_site());
```

kennytm commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,591 @@ impl RpnFunctionOpts {
+        generics
+            .params
+            .push(parse_str(&format!("D_: {}", ARG_DEF)).unwrap());
+        let fn_name = format!("{}", &self.item_fn.ident);
```
(to_string)

kennytm commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,591 @@ impl RpnFunctionOpts {
+            .push(parse_str(&format!("D_: {}", ARG_DEF)).unwrap());
+        let fn_name = format!("{}", &self.item_fn.ident);
+        let fn_trait_ident = &self.fn_trait_ident;
+        let tp_ident = parse_str::<Ident>("D_").unwrap();
```
(Ident::new)

kennytm commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,591 @@ impl RpnFunctionOpts {
+        //            call_args += &format!("{}, ", arg_var);
+        //        }
+        //        let extract_args: TokenStream = extract_args.parse().unwrap();
+        //        let call_args: TokenStream = call_args.parse().unwrap();
```
What is this block of commented code?

kennytm commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,591 @@ impl RpnFunctionOpts {
+        //        let call_args: TokenStream = call_args.parse().unwrap();
+        let meta = &self.meta;
+        let extract = (0..self.arg_types.len())
+            .map(|i| syn::parse_str::<Ident>(&format!("arg{}", i)).unwrap());
```
(Ident::new)

kennytm commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,591 @@ impl RpnFunctionOpts {
+                    let mut result = Vec::with_capacity(rows);
+                    for row in 0..rows {
+                        #(let (#extract, arg) = arg.extract(row));*;
+                        result.push( #fn_ident #ty_generics_turbofish ( #(#meta),* #(#call_arg),* )?);
```
This will expand to `(meta1, meta2 call_arg1, call_arg2)`, missing the comma between the `meta`s and `call_arg`s.

sticnarf replied:
Oh, I see. I think it's ok to write `#(#meta,)* #(#call_arg),*`

kennytm commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,579 @@ impl RpnFunctionOpts {
+const ARG_DEF: &str = "crate::coprocessor::dag::rpn_expr::function::ArgDef";
+const NULL: &str = "crate::coprocessor::dag::rpn_expr::function::Null";
+const RPN_FN_ARG: &str = "crate::coprocessor::dag::rpn_expr::function::RpnFnArg";
+const ARG: &str = "crate::coprocessor::dag::rpn_expr::function::Arg";
```
Could these four turn into something like `ctx_type()` below?

lonng replied:
I don't think we really need these constants.

kennytm commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,579 @@ impl RpnFunctionOpts {
+        } else {
+            let types = format!("{}", ty_generics.clone().into_token_stream());
+            let phantom_generic: TokenStream =
+                parse_str(&format!("<({})>", &types[1..types.len() - 1])).unwrap();
```
```rust
let types_middle = &types[1..(types.len() - 1)];
let phantom_generic = quote!{ <(#(#types_middle),*)> };
```

sticnarf replied:
`types` here is a string, we cannot write this way...I decide to use `generics.type_params().map(|type_param| type_param.ident)`, and put it in the `quote!`. Lifetime won't appear here.

kennytm commented on components/cop_codegen/src/rpn_function.rs:
```diff
@@ -62,3 +71,579 @@ impl RpnFunctionOpts {
+        for arg_index in 0..arg_len {
+            evaluator = format!("ArgConstructor::new({}, {})", arg_index, evaluator);
+        }
+        let evaluator: TokenStream = evaluator.parse().unwrap();
```
You don't need to stringify at all?

```rust
let evaluator_ident = &self.evaluator_ident;
let mut evaluator = if self.item_fn.decl.generics.params.empty_or_trailing() {
    quote!{ #evaluator_ident }
} else {
    quote!{ #evaluator_ident #ty_generics_turbofish (std::marker::PhantomData) }
};
let arg_len = self.arg_types.len();
for arg_index in 0..arg_len {
    evaluator = quote!{ ArgConstructor::new(#arg_index, #evaluator) };
}
```

kennytm commented on src/coprocessor/dag/rpn_expr/function.rs:
```diff
@@ -345,3 +423,171 @@ impl Helper {
+pub trait Evaluator {
+    fn eval<D: ArgDef>(
+        self,
+        def: D,
```
You could omit all those `D_: ArgDef` in the proc macro if you use `def: impl ArgDef` here 🙃

sticnarf commented:
The latest commit https://github.com/tikv/tikv/pull/4860/commits/d660f0d1f9cc58207d80bf8adcb9b53c92e67958 addresses all issues mentioned above. 
PTAL again. @kennytm @lonng Thanks!

sticnarf commented:
/run-integration-tests

breeswish commented on src/coprocessor/dag/rpn_expr/function.rs:
````diff
@@ -1,5 +1,83 @@
+//! understand how `Evaluator` and `RpnDef` work. There's a procedure macro called `rpn_fn`
+//! helping you create RPN functions. For example:
+//!
+//! ```
````
Will this cause issues when running doc tests?

sticnarf replied:
Probably...Let me add the `no_run` attribute.

sticnarf commented:
/run-integration-tests

lonng commented on src/coprocessor/dag/rpn_expr/function.rs:
````diff
@@ -1,5 +1,83 @@
+//! ```
+//!
+//! If you are curious about what code the macro will generate, check the test code
+//! in `components/cop_codegen/rpn_functions.rs`.
````
```suggestion
//! in `components/cop_codegen/src/rpn_function.rs`.
```

sticnarf commented:
/run-integration-tests
